explainer.md

```diff
@@ -23,7 +23,22 @@ const session = await navigator.xr.requestSession(“immersive-ar”, {
 });
 ```
 
-2. In requestAnimationFrame callback, query XRFrame for the currently available XRDepthInformation:
+2. Configure the session to set the depth sensing usage and data format.
+
+The application can query the user agent's preferred usage (`"cpu-optimized"` or `"gpu-optimized"`) and format (`"luminance-alpha"` or `"float32"`) by accessing `session.preferredDepthSensingUsage` & `session.preferredDepthFormat`:
```

Comment:
We need to already know the device that's going to be used in order to provide the preferred settings. For example, different format may be preferred for "immersive-vr" session than for "immersive-ar" session. Additionally, I was also not sure if exposing this information even before the user gave any consent is the right way to go (fingerprintability may be an issue here).

Reply:
Would it make sense to drop the preferred usage and format?
What if an author could provide a choice of flags? If one flag is passed, the UA has to handle that but if both are passed, the UA is free to choose the optimal one from the list.

I suspect that very few experiences will switch between modes and this will make the API much simpler.

Reply:
It's not only about switching between modes (even though I don't see how we could make it work) - even assuming that the site requests only "immersive-vr" session, the UA would need to provide a value depending on what is the currently connected hardware or framework it needs to use to communicate with it (even though the API may not even be supported at all - on Android w/ ARCore we only know if depth API is supported when we try to create ARCore's session).

> What if an author could provide a choice of flags? If one flag is passed, the UA has to handle that but if both are passed, the UA is free to choose the optimal one from the list.

Do you mean that the app would pass in an array in the order of decreased preferences for what it wants to use, and the UA will pick something & communicate to the app what was selected? I think that could work as well, but I don't think we should force the UA to implement support for all possible combinations of values (i.e. if the UA doesn't see something it supports on the list, it will reject the session). Then we won't need the configuration step, but we will need to expose the usage and format that was selected.

Reply:
> It's not only about switching between modes (even though I don't see how we could make it work) - even assuming that the site requests only "immersive-vr" session, the UA would need to provide a value depending on what is the currently connected hardware or framework it needs to use to communicate with it (even though the API may not even be supported at all - on Android w/ ARCore we only know if depth API is supported when we try to create ARCore's session).

By "modes", I meant different types of depth sensing modes: cpu vs gpu, not the type of session

> > What if an author could provide a choice of flags? If one flag is passed, the UA has to handle that but if both are passed, the UA is free to choose the optimal one from the list.
> 
> Do you mean that the app would pass in an array in the order of decreased preferences for what it wants to use, and the UA will pick something & communicate to the app what was selected? 

Yes
>I think that could work as well, but I don't think we should force the UA to implement support for all possible combinations of values (i.e. if the UA doesn't see something it supports on the list, it will reject the session). Then we won't need the configuration step, but we will need to expose the usage and format that was selected.

Correct. I don't really like that a UA doesn't have to provide support for all approaches as that lead to messy code where authors might try to polyfill the missing mode but we can deal with that separately.

Reply:
> (...) but we can deal with that separately.

Agreed, let's focus on the API shape first and we can iron out all the wrinkles in the spec text once we know roughly what we want from the API.

I quite like the approach with the app providing the lists of preferences, but I do see one problem with it: it'd be tricky for the app to express all different combinations of parameters (i.e. w/ one list for usage preferences and one list for data format preferences, the app won't be able to say "I prefer gpu over cpu; for gpu-optimized usage, I prefer luminance-alpha data format; for cpu-optimized usage, I prefer float32 data format"). If we add to that the requirement that the UA must support everything, the app may end up w/ something it doesn't really prefer.

The "I prefer gpu over cpu; for gpu-optimized usage, I prefer luminance-alpha data format; for cpu-optimized usage, I prefer float32 data format" example preference could be expressed as 2 session requests that could be issued:
1. ["gpu-optimized"], ["luminance-alpha", "float32"], but if that fails, I'll try w/ a 2nd request
2. ["cpu-optimized"], ["float32", "luminance-alpha"]

Reply:
Resolving, please take a look at the latest version and reopen if I misunderstood something here.